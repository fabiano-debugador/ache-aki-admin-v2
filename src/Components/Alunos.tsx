
import React, { useEffect, useState } from "react";
import { Table, Button, Form } from 'react-bootstrap';
import axios, { AxiosResponse } from "axios";

type Aluno = {
    id?: string | null;
    nome: string;
    telefone: string;
    pix: string;
};

type SubmitFormProps = {
    preventDefault: () => void; 
}

export function Alunos()  {
    const initialState = { id: null, nome: "", telefone: "", pix: "" }
    const [alunos , setAlunos] = useState<Aluno[]>([]);
    const [aluno, setAluno] = useState<Aluno>(initialState);

    function handleInput (e: React.ChangeEvent<HTMLInputElement>): void {
        const newObj = { ...aluno };
        const { name, value } = e.target;
        setAluno({ ...newObj, [name]: value })
    }

    function create(body: Aluno): void {
        let id = 'id' + (new Date()).getTime();
        const data: Aluno = { ...body, id }
        axios.post('http://localhost:3004/profile', data)
        setAlunos([ ...alunos, data]);
        setAluno(initialState);
    }

    function update(body: Aluno) {
        var index = alunos.map(function(e: Aluno) { return e.id; }).indexOf(body.id);
        alunos[index].nome = body.nome;
        alunos[index].telefone = body.telefone;
        alunos[index].pix = body.pix;
        axios.put('http://localhost:3004/profile/'+aluno.id, body);
        setAlunos([...alunos]);
        setAluno(initialState);
    }

    function getAluno(id: string): void {
        const selectedAluno = alunos.filter((aluno: Aluno) => {return aluno.id === id });
        setAluno(selectedAluno[0])
    }

    function deletar(id: string): void {
        try {
            axios.delete('http://localhost:3004/profile/'+id)
            const listaDeAlunos = alunos.filter((aluno: Aluno) => {return aluno.id !== id })
            setAlunos(listaDeAlunos)
        } catch(e) {
            console.log(e)
        }
    }

    function getUserAccount(): Promise<AxiosResponse<Aluno>> | undefined {
        try{
           return axios.get('http://localhost:3004/profile') 
            
        }catch(e: any) {
            console.log(e)
        }
      }

      useEffect(() => {
       getUserAccount()?.then(function (response: any) {
        setAlunos(response.data)
      })
      .catch(function (error: Error) {
        console.log(error);
      });
      },[])
   
      const submitForm = (event: SubmitFormProps): void => {
        event.preventDefault()

        if (aluno.id) {
            update(aluno)
        } else {
            create(aluno)
        }
      }
  
        return (
            <div>
                <Form onSubmit={submitForm}>
                    <Form.Group className="mb-3" controlId="formBasicEmail">
                        <Form.Label>Nome</Form.Label>
                        <Form.Control type="text" name="nome" value={aluno.nome} onChange={(e: React.ChangeEvent<HTMLInputElement>) => handleInput(e)} placeholder="Nome" />
                    </Form.Group>

                    <Form.Group className="mb-3" controlId="formBasicPassword">
                        <Form.Label>Telefone</Form.Label>
                        <Form.Control type="text" name="telefone" value={aluno.telefone} onChange={(e: React.ChangeEvent<HTMLInputElement>) => handleInput(e)} placeholder="Telefone" />
                    </Form.Group>
                    <Form.Group className="mb-3" controlId="formBasicPassword">
                        <Form.Label>Pix</Form.Label>
                        <Form.Control type="text" name="pix" value={aluno.pix} onChange={(e: React.ChangeEvent<HTMLInputElement>) => handleInput(e)} placeholder="Pix" />
                    </Form.Group>
                    <Button variant="primary" type="submit">
                        Submit
                    </Button>
                </Form>
                <Table striped bordered hover>
                    <thead>
                        <tr>
                            <td>Nome</td>
                            <td>Telefone </td>
                            <td>Pagamento via Pix </td>
                            <td> Op√ßoes </td>
                        </tr>
                    </thead>

                    <tbody>
                        {
                            alunos.map((aluno: any) =>
                                <tr key={aluno.id}>
                                    <td>{aluno.nome}</td>
                                    <td>{aluno.telefone}</td>
                                    <td> {aluno.pix} </td>
                                    {aluno.id && (
                                        <td>
                                            <Button variant="danger" onClick={() => getAluno(aluno?.id)}>Atualizar</Button> / 
                                            <Button variant="danger" onClick={() => deletar(aluno?.id)}>Deletar</Button>
                                        </td>
                                    )}
                                </tr>
                            )
                        }
                    </tbody>
                </Table>

            </div>
        )
    
}

