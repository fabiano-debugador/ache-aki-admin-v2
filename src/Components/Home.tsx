import React from 'react';

export function Home() {
  return (
    <div>
      <p>Pagina do Home </p>
    </div>
  );
}
