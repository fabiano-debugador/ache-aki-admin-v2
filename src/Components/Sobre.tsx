import React from 'react';
export function Sobre() {
  return (
    <div>
      <p>Sobre os Alunos </p>
    </div>
  );
}
