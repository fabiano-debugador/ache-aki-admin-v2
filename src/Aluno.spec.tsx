import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';
import { Alunos } from './Components/Alunos';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';

describe('fetchUsers', () => {
  let mock: MockAdapter;

  beforeAll(() => {
    mock = new MockAdapter(axios);
  });

  afterEach(() => {
    mock.reset();
  });

  describe('Testing CRUD', () => {
    const studants = [
      {
        id: 'id1676981397699',
        nome: 'mario',
        telefone: '11 90797979',
        pix: '7979/0900-98',
      },
      {
        id: 'id1677083499896',
        nome: 'Valentino Rossi',
        telefone: '',
        pix: '',
      },
      {
        id: 'id1677102891363',
        nome: 'Jéssica',
        telefone: '19 9098098080',
        pix: 'jessica@mail.com',
      },
    ];
    it('load page', async () => {
      mock.onGet(`http://localhost:3004/profile`).reply(200, studants);

      const { findAllByTestId } = render(<Alunos />);

      const text = await findAllByTestId('listOfStudants');

      expect(text).toHaveLength(3);
    });

    it('create studant', async () => {
      const studants = [
        {
          id: 'id1676981397699',
          nome: 'mario',
          telefone: '11 90797979',
          pix: '7979/0900-98',
        },
        {
          id: 'id1677083499896',
          nome: 'Valentino Rossi',
          telefone: '',
          pix: '',
        },
        {
          id: 'id1677102891363',
          nome: 'Jéssica',
          telefone: '19 9098098080',
          pix: 'jessica@mail.com',
        },
      ];

      mock.onGet(`http://localhost:3004/profile`).reply(200, studants);

      const { findAllByTestId, getByTestId } = render(<Alunos />);
      const fieldName = getByTestId('nome');
      const submit = getByTestId('submit-button');
      let allStudants = await findAllByTestId('listOfStudants');

      expect(allStudants).toHaveLength(3);

      fireEvent.change(fieldName, { target: { value: 'ruberley' } });

      mock.onPost('http://localhost:3004/profile').reply(200);

      fireEvent.click(submit);

      studants.push({
        id: 'id1677102891674',
        nome: 'ruberley',
        telefone: '19982266776',
        pix: '83888337367',
      });

      allStudants = await findAllByTestId('listOfStudants');

      expect(allStudants).toHaveLength(4);
    });

    it('Update studant', async () => {
      const studants = [
        {
          id: 'id1676981397699',
          nome: 'mario',
          telefone: '11 90797979',
          pix: '7979/0900-98',
        },
        {
          id: 'id1677083499896',
          nome: 'Valentino Rossi',
          telefone: '',
          pix: '',
        },
        {
          id: 'id1677102891363',
          nome: 'Jéssica',
          telefone: '19 9098098080',
          pix: 'jessica@mail.com',
        },
      ];

      mock.onGet(`http://localhost:3004/profile`).reply(200, studants);

      const { findAllByTestId, getByTestId } = render(<Alunos />);
      const fieldName = getByTestId('nome');
      const submit = getByTestId('submit-button');
      const buttonUpdate = await screen.findAllByTestId('update');
      let allStudants = await findAllByTestId('listOfStudants');

      expect(allStudants).toHaveLength(3);

      fireEvent.click(buttonUpdate[0]);
      fireEvent.change(fieldName, { target: { value: 'mario e luigi' } });

      mock.onPut('http://localhost:3004/profile/id1676981397699').reply((config) => {
        expect(JSON.parse(config.data)).toEqual({
          id: 'id1676981397699',
          nome: 'mario e luigi',
          telefone: '11 90797979',
          pix: '7979/0900-98',
        });
        return [200];
      });

      fireEvent.click(submit);

      allStudants = await findAllByTestId('listOfStudants');

      expect(allStudants[0]).toMatchSnapshot();
    });

    it('Delete studant', async () => {
      const studants = [
        {
          id: 'id1676981397699',
          nome: 'mario',
          telefone: '11 90797979',
          pix: '7979/0900-98',
        },
        {
          id: 'id1677083499896',
          nome: 'Valentino Rossi',
          telefone: '',
          pix: '',
        },
        {
          id: 'id1677102891363',
          nome: 'Jéssica',
          telefone: '19 9098098080',
          pix: 'jessica@mail.com',
        },
      ];

      mock.onGet(`http://localhost:3004/profile`).reply(200, studants);

      const { findAllByTestId } = render(<Alunos />);
      const deleteButton = await screen.findAllByTestId('delete');
      let allStudants = await findAllByTestId('listOfStudants');

      expect(allStudants).toHaveLength(3);

      mock.onDelete('http://localhost:3004/profile/id1676981397699').reply(204);

      fireEvent.click(deleteButton[0]);

      allStudants = await findAllByTestId('listOfStudants');

      expect(allStudants).toHaveLength(2);
    });
  });
});
