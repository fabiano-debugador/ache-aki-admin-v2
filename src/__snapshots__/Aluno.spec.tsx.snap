// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchUsers Delete studant 1`] = `
[
  <tr
    data-testid="listOfStudants"
  >
    <td>
      Valentino Rossi
    </td>
    <td />
    <td>
       
       
    </td>
    <td>
      <button
        class="btn btn-danger"
        data-testid="update"
        type="button"
      >
        Atualizar
      </button>
       
      /
      <button
        class="btn btn-danger"
        data-testid="delete"
        type="button"
      >
        Deletar
      </button>
    </td>
  </tr>,
  <tr
    data-testid="listOfStudants"
  >
    <td>
      Jéssica
    </td>
    <td>
      19 9098098080
    </td>
    <td>
       
      jessica@mail.com
       
    </td>
    <td>
      <button
        class="btn btn-danger"
        data-testid="update"
        type="button"
      >
        Atualizar
      </button>
       
      /
      <button
        class="btn btn-danger"
        data-testid="delete"
        type="button"
      >
        Deletar
      </button>
    </td>
  </tr>,
]
`;

exports[`fetchUsers Testing CRUD Delete studant 1`] = `
[
  <tr
    data-testid="listOfStudants"
  >
    <td>
      Valentino Rossi
    </td>
    <td />
    <td>
       
       
    </td>
    <td>
      <button
        class="btn btn-danger"
        data-testid="update"
        type="button"
      >
        Atualizar
      </button>
       
      /
      <button
        class="btn btn-danger"
        data-testid="delete"
        type="button"
      >
        Deletar
      </button>
    </td>
  </tr>,
  <tr
    data-testid="listOfStudants"
  >
    <td>
      Jéssica
    </td>
    <td>
      19 9098098080
    </td>
    <td>
       
      jessica@mail.com
       
    </td>
    <td>
      <button
        class="btn btn-danger"
        data-testid="update"
        type="button"
      >
        Atualizar
      </button>
       
      /
      <button
        class="btn btn-danger"
        data-testid="delete"
        type="button"
      >
        Deletar
      </button>
    </td>
  </tr>,
]
`;

exports[`fetchUsers Testing CRUD Update studant 1`] = `
<tr
  data-testid="listOfStudants"
>
  <td>
    mario e luigi
  </td>
  <td>
    11 90797979
  </td>
  <td>
     
    7979/0900-98
     
  </td>
  <td>
    <button
      class="btn btn-danger"
      data-testid="update"
      type="button"
    >
      Atualizar
    </button>
     
    /
    <button
      class="btn btn-danger"
      data-testid="delete"
      type="button"
    >
      Deletar
    </button>
  </td>
</tr>
`;
